Handshaking

annym opened the pull request:
## Overview

Description:

An implementation of Corfu client-server handshaking. Overall:
- Upon connection, the client will initiate a handshake.
- Both (server and client) will start a handshake timeout timer.
- The client sends a message that contains its own id and the (asserted) server's id.
- The server validates the handshake message and replies with its own server id
  and corfu version.
- The client makes its own validation and in case of is success,
  message exchange is enabled.
- If any validation fails client/server will close the connection.


Why should this be merged:  This is a Feature Request. The fact that a node is no longer tied to an IP and instead is uniquely identified by a 16Byte Identifier raises the need of actually validating upon connection that we are connected to the right node. 

Related issue(s) (if applicable): #1077 


## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

no2chem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,213 @@
+                ctx.channel().remoteAddress());
+
+        // Start timeout checker. Wait for handshake to complete, otherwise close connection.
+        new Thread() {
```
Hm. I think we want to avoid creating a new thread here. I think we have two options here, we can schedule on the executor for the context (```ctx.executor()```), or we can use the ```ReadTimeoutHandler``` and uninstall it from the pipeline as well (I think you can uninstall by name). I'm sort of leaning towards the ReadTimeoutHandler.

annym replied:
I wasn't aware of the ReadTimeoutHandler, it is really useful. I also lean towards this approach. What I am doing now is adding the ReadTimeoutHandler upon handshake initiation, and removing it on ChannelRead.. This also removes the latch and the mutex

no2chem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,213 @@
+                    latch.await(timeoutInMillis, TimeUnit.MILLISECONDS);
+                } catch (InterruptedException e) {
+                    log.error("Handshake timeout checker: interrupted!");
+                    e.printStackTrace();
```
Should re-throw the exception :).

no2chem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,213 @@
+    private final UUID nodeId;
+    private final String corfuVersion;
+    private final ChannelGroup group;
+    private final AtomicBoolean handshakeComplete;
```
volatile boolean should be sufficient, right?

no2chem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,213 @@
+ * Created by amartinezman on 12/11/17.
+ */
+@Slf4j
+@ChannelHandler.Sharable
```
Is this really Sharable? handshakeComplete looks like state.

annym replied:
Totally agree, I thought I had removed it. It is not sharable.

no2chem commented on runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java:
```diff
@@ -0,0 +1,223 @@
+        ChannelFuture f = ctx.writeAndFlush(handshake);
+        f.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future)
```
Same re: ReadTimeoutHandler on the server side.

annym replied:
Done

no2chem commented on runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:
```diff
@@ -422,7 +422,7 @@ public static UUID getCheckpointStreamIdFromName(String streamName) {
 
     /**
-     * Get corfy runtime version.
+     * Get corfu runtime version.
```
Heh, good catch

annym commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,213 @@
+            try {
+                handshake = (CorfuPayloadMsg<HandshakeMsg>) m;
+            } catch (ClassCastException e) {
+                log.warn("Non-handshake message received by handshake handler. Send upstream only " +
```
Done

annym commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,213 @@
+            }
+
+            // Store clientID as a channel attribute.
+            AttributeKey<UUID> clientIdAttrKey = AttributeKey.valueOf("ClientID");
```
Done

annym commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,213 @@
+            // Store clientID as a channel attribute.
+            AttributeKey<UUID> clientIdAttrKey = AttributeKey.valueOf("ClientID");
+            ctx.channel().attr(clientIdAttrKey).set(clientId);
+            log.info("Handshake validated by Server.");
```
Done

annym commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,213 @@
+            }
+
+            // Remove this handler from the pipeline; handshake is completed.
+            log.info("Removing handshake handler from pipeline.");
```
Done

annym commented:
@no2chem I have addressed your comments.  Please let me know if there is any other comment. The only thing remaining is replacing to the actual NodeId, so there is a dependency on these other PRs.

no2chem commented on generator/src/main/java/org/corfudb/generator/LongevityApp.java:
```diff
@@ -114,17 +114,17 @@ private void waitForAppToFinish() {
-
-            exitStatus = checkpointHasFinished ? 0 : 1;
-            System.exit(exitStatus);
+//            boolean checkpointHasFinished = false;
```
Is there a reason this is part of this PR?

annym replied:
oops.. will remove, this was part of another test I was doing... plus I'll add the javadocs

medhavidhawan commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,197 @@
+        log.info("channelActive: Incoming connection established from: " +
+                ctx.channel().remoteAddress() + " Start Read Timeout.");
+        ctx.pipeline().addBefore(ctx.name(), "readTimeoutHandler",
+                new ReadTimeoutHandler(10));
```
should 10 be a constant or a config property ?

no2chem replied:
Hm, yeah it should be a config property. Can you pull this from the options map in ServerContext?

annym replied:
Done

medhavidhawan commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+
+    private void fireHandshakeFailed(ChannelHandlerContext ctx) {
```
do we need 2 variables "handshakeComplete and handshakeFailed" to maintain state ?

annym replied:
I believe we should use 2. The fact that handshake hasn't failed (default) does not imply it has completed. Only if it has completed we might forward incoming messages in the pipeline when received in the window of handshake validation when handler has not been removed yet from the pipeline.

medhavidhawan replied:
The variables you use are java objects. These variables have a derived state "initialized/null" and you can use that for the signal you are looking for. I am ok with your approach and do not have a strong opinion. If 2 variables make it easier for you, I would suggest to couple them together in a state object so that they mutate together and there is little chance of making mistakes where we set one and forgot the other. The code that you have does not have any issues, the logic looks good , this is more of suggestion for future proofing.

annym commented on runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java:
```diff
@@ -0,0 +1,197 @@
+           handshakeResponse = (CorfuPayloadMsg<HandshakeResponse>) m;
+            log.info("channelRead: Handshake Response received. Removing readTimeoutHandler " +
+                    "from pipeline.");
+            ctx.pipeline().remove("readTimeoutHandler");
```
Done

no2chem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object m) throws Exception {
```
I think for once this is rebased against master with the CorfuServer branch, integration testing will be possible. Unit testing will have to unfortunately come a little later.

medhavidhawan replied:
ok cool.

medhavidhawan replied:
ok cool

annym commented on runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java:
```diff
@@ -0,0 +1,197 @@
+
+    private final UUID clientId;
+    private final UUID nodeId;
+    private final AtomicBoolean handshakeComplete;
```
As before, you might enter a state in which handshake hasn't failed (the default value) but not because it actually completed but because nothing has happened yet. Or it might have failed not because it completed but because it timed out.

medhavidhawan replied:
So the variable you use is an object. It has a third "construed" state: uninitialized which can be used for that, But as long as you are comfortable, I am ok. My suggestion would be to couple these variables together in one object so that they mutate together and there is less chance of mistakes for later as well.

annym commented on runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object m)
```
I have included comments on all methods. Yes, we do plan to have integration tests.

annym commented on runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java:
```diff
@@ -0,0 +1,197 @@
+        } catch (ClassCastException e) {
+            log.warn("channelRead: Non-handshake message received by handshake handler. Send upstream only " +
+                    "if handshake succeeded.");
+            if (this.handshakeComplete.get()) {
```
Done

annym commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object m) throws Exception {
```
Done

Maithem commented:
The commit history is a bit off, please rebase.

no2chem commented:
@annym 
It looks like its failing because the option wasn't set on the TestServerContext
```
                    context.getServerConfig(Integer.class, "--HandshakeTimeout")));
```
is null. 

In the test package :  org.corfudb.infrastructure.ServerContextBuilder
Create a new handshakeTimeout field and set that field in .ServerContextBuilder::Build

no2chem commented:
@annym Also might want to double check that you get an integer (seems like docopt will probably actually give you a string).

annym commented:
@medhavidhawan I get your point. I have created a Handshake State Object so both mutate together. 

@no2chem Good catch, already added it to ServerContextBuilder as a String and made the required changes.

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **aed31d9**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1078 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/sGKfPzApcJNo454tOh58vCKMWpkRSlwS)
